### PR TITLE
stacktrace: skip frames

### LINKED ISSUE #512: CallerSkip does not apply to stacktraces captured by zap
Currently, logging wrappers use `CallerSkip` to avoid the logging wrapper showing up as the caller. However, this is not used by stacktraces, so we end up with the logging wrappers showing up in stacktraces.

@akshayjshah: I'm thinking we should respect caller skip in stacktraces as well, any thoughts?

### PR DESCRIPTION
**Pull request by boz**
* Honor StackSkip option when including stacktrace.
 * Replace package-based stack pruning - the same behavior is achieved by
   skipping frames appropriately.

 fixes #512

### CODE REVIEW

**Comment by boz** [comment]
Hi @prashantv, @akshayjshah.  Any thoughts on this?

Thanks!

File: field.go [hunk 1]
```diff
@@ -184,11 +184,19 @@ func Time(key string, val time.Time) zapcore.Field {
...
 // takes about two microseconds.
 func Stack(key string) zapcore.Field {
+	return String(key, takeStacktrace(1))
+}
+
+// StackSkip returns a stacktrace field like Stack and also allows for skipping
+// stack frames via the skip argument.
+//
+// StackSkip("foo",0) is equivalent to Stack("foo").
+func StackSkip(key string, skip int) zapcore.Field {
```
Comment (prashantv) [inline field.go]: I'm not sure if we should expose this right now -- I'd prefer to keep just `Stack` for now and think about whether we really need the skip version separately.

File: stacktrace.go [hunk 2]
```diff
@@ -22,38 +22,31 @@ package zap
...
+func takeStacktrace(skip int) string {
 	buffer := bufferpool.Get()
 	defer buffer.Free()
 	programCounters := _stacktracePool.Get().(*programCounters)
 	defer _stacktracePool.Put(programCounters)
 
 	var numFrames int
+	skip += 2
 	for {
 		// Skip the call to runtime.Counters and takeStacktrace so that the
```
Comment (prashantv) [inline stacktrace.go]: can you move this comment to where we do `+= 2`

File: stacktrace_ext_test.go [hunk 3]
```diff
@@ -37,22 +32,23 @@ import (
...
 		logger.Error("test log")
 		logger.Sugar().Error("sugar test log")
 
-		require.Contains(t, out.String(), "TestStacktraceFiltersZapLog", "Should not strip out non-zap import")
-		verifyNoZap(t, out.String())
+		frames := strings.Split(out.String(), "\n")
+
+		assert.True(t, len(frames) > 6)
+
+		// The logging function should be the first frame ...
```
Comment (prashantv) [inline stacktrace_ext_test.go]: instead of testing partially, can we just test the entire result of `frames`?

File: stacktrace.go [hunk 4]
```diff
@@ -63,19 +56,12 @@ func takeStacktrace() string {
 	}
 
 	i := 0
-	skipZapFrames := true // skip all consecutive zap frames at the beginning.
```
Comment (prashantv) [inline stacktrace.go]: This behaviour was added on purpose, any reason why we're deleting this logic?

**Comment by prashantv** [comment]
Thanks for the contribution @boz 

Can you provide more context about why we're removing the `zap` stacktrace pruning?

**Comment by boz** [comment]
Hi @prashantv 

As far as I can tell, accounting for frames correctly with `skip` and `zap.StackSkip()` achieves the same thing that the name-based frame skipping did.  I might be missing something, tho.

File: stacktrace_ext_test.go (hunk 3 above)
Comment (boz) [reply stacktrace_ext_test.go]: Sure, added.  I was worried it'd be too brittle.

**Comment by akshayjshah** [comment]
Today, the offset for determining callers is separate from the stack-pruning behavior. That means that it's entirely possible to use a logger that *never* includes the calling function, but does occasionally log stacktraces. In that case, the caller skip likely wouldn't be configured, but stacktraces would still prune zap's internals.

Unfortunately, that means that this change isn't backward-compatible and can't ship as-is :/

In retrospect, I regret the offset-based skip mechanism we used to determine the calling function. It's useful, but has turned out to be even more fiddly and error-prone than I expected.

**Comment by akshayjshah** [comment]
This has been open for a long time, and I'd prefer not to introduce even small backward-incompatible changes if they're not absolutely necessary. For housekeeping, I'm going to close this PR - @boz, happy to discuss further if you want to re-open it.

**Pull request closed without merging**